tests/utils/testapp/gen-syscall-events-callstack/gen-syscall-events-callstack.c

```diff
@@ -0,0 +1,154 @@
+/*
+ * Copyright (C) - 2017 Francis Deslauriers <francis.deslauriers@efficios.com>
+ *
+ * This library is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as published by the
+ * Free Software Foundation; version 2.1 of the License.
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public License
+ * for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this library; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301 USA
+ */
+
+#include <fcntl.h>
+#include <signal.h>
+#include <stdbool.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/syscall.h>
+#include <sys/types.h>
+#include <unistd.h>
+
+/**
+ * Registers a signal handler and waits for a signal before executing a syscall
+ * and exiting. This is useful for tests in combinaison with LTTng's PID tracker
+ * feature where we can trace the kernel events generated by our test process
+ * only.
+ */
+
+volatile int ready = 0;
+volatile int val = 0;
+
+void sig_handler(int signum)
+{
+	ready = 1;
+}
+
+int create_file(const char *path)
+{
+	static bool file_created = false;
+	int ret;
+	if (!path || file_created) {
+		return -1;
+	}
+
+	ret = creat(path, S_IRUSR | S_IWUSR);
+	if (ret < 0) {
+		fprintf(stderr, "Failed to create file %s\n", path);
+		return -1;
+	}
+
+	(void) close(ret);
+	file_created = true;
+
+	return 0;
+}
+
+long __attribute__ ((noinline))
+my_gettid(void)
+{
+    long ret;
+    asm volatile
+    (
+        "syscall"
+        : "=a" (ret)
+        : "0"(__NR_gettid)
+        : "cc", "rcx", "r11", "memory"
+    );
+    return ret;
+}
+
+int __attribute__ ((noinline))
+fct_c(void)
+{
+	my_gettid();
+	return rand() % 42;
+}
+
+int __attribute__ ((noinline))
+fct_b(void)
+{
+	val += fct_c();
+	return val;
+}
+
+int __attribute__ ((noinline))
+fct_a(void)
+{
+	val += fct_b();
+	return val;
+}
+
+int main(int argc, char **argv)
+{
+	int ret = 0;
+	char *path;
+	struct sigaction sig_action;
+
+	if (argc != 2) {
+		fprintf(stderr, "Error: Missing argument\n");
+		fprintf(stderr, "USAGE: %s PATH_SYNC_FILE\n", argv[0]);
+		ret = -1;
+		goto error;
+	}
+
+	path=argv[1];
+
+	/**
+	 * Register a signal handler, create a file to signify that the signal
+	 * handler is registered and wait for a signal from the parent process
+	 * to go on with the function call.
+	 */
+	sig_action.sa_handler = sig_handler;
+
+	ret = sigemptyset(&sig_action.sa_mask);
+	if (ret == -1) {
+		perror("sigemptyset");
+		goto error;
+	}
+
+	ret = sigaction(SIGUSR1, &sig_action, NULL);
+
+	if (ret == -1) {
+		perror("sigaction");
+		goto error;
+	}
+
+	/* Create a file to signify that the signal handler is registered */
```

Comment:
you should look more into the existing test apps for synchronization. The points at which the rendez-vous points are placed are important, and there are race scenarios that are not covered here.